File: p208.cpp
#include <bits/stdc++.h>

using namespace std;

class TrieNode {    
public:
    bool is_end;
    TrieNode* children[26];
    
    TrieNode() {
        for(int i = 0; i < 26; ++ i) {
            children[i] = NULL;
        }
        is_end = false;
    }
};

class Trie {
private:
    TrieNode* head;
public:
    /** Initialize your data structure here. */
    Trie() {
        head  = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode* curr = head;
        for(char c : word) {
            if(curr->children[c-'a'] == NULL)
                curr->children[c-'a'] = new TrieNode();
            curr = curr->children[c-'a'];
        }
        curr->is_end = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        TrieNode* curr = head;
        for(char c : word) {
            if(curr->children[c-'a'] == NULL) {
                return false;
            }
            curr = curr->children[c-'a'];
        }
        return curr->is_end;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        TrieNode* curr = head;
        for(char c : prefix) {
            if(curr->children[c-'a'] == NULL) {
                return false;
            }
            curr = curr->children[c-'a'];
        }
        return true;
    }
};

File: p208.py
class TrieNode:

    def __init__(self, val, end):
        self.val = val
        self.is_end = end
        self.children = [None]*26

class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.head = TrieNode('', False)

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        curr = self.head
        for i, c in enumerate(word):
            if curr.children[ord(c)-ord('a')] is None:
                curr.children[ord(c)-ord('a')] = TrieNode(c, False)
            if i == len(word) - 1:
                curr.children[ord(c)-ord('a')].is_end = True
            curr = curr.children[ord(c)-ord('a')]

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        curr = self.head
        for i,c in enumerate(word):
            if curr.children[ord(c)-ord('a')] is None:
                return False
            curr = curr.children[ord(c)-ord('a')]
        return curr.is_end

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        curr = self.head
        for i,c in enumerate(prefix):
            if curr.children[ord(c)-ord('a')] is None:
                return False
            curr = curr.children[ord(c)-ord('a')]
        return True

